### PR TITLE
Show first-introduced versions on roadmap

### LINKED ISSUE #187: Please change roadmap checkmark chart to caniuse style "first introduced in"
The page https://webassembly.org/roadmap/ is a great improvement in gauging the different feature sets of Wasm. Would it be possible to change that chart to instead of checkmarks, show the version of the browser where the feature was first introduced, like caniuse.com works?

That works better when one has to make decisions that balance between supporting old browsers vs when it's safe to assume that a certain feature is ubiquitous.

### PR DESCRIPTION
**Pull request by andylizi**
This is a tentative implementation of the long-awaited feature proposed in #187.

feature-table

The initial data set is from [this comment](https://github.com/WebAssembly/website/issues/187#issuecomment-1222992930).

Closes #187.

### CODE REVIEW

**Comment by juj** [comment]
This looks *absolutely* **fantastic**!

**Comment by RReverser** [comment]
This does look amazing! I can't do a proper review at the moment, could you ping me in a 3-4 days in case I forget? (others are welcome to review meanwhile, too)

**Comment by andylizi** [comment]
> I can't do a proper review at the moment, could you ping me in a 3-4 days in case I forget?

@RReverser ping~

**Comment by RReverser** [comment]
Thanks, I built and played with this locally, and overall I like it a lot. Two notes:

1. In previous version table cells with notes were focusable, which provided at least some primitive level of accessibility for keyboard users (although not a great one). Now that's replaced by [a] letters which also help to relate cells to footnotes at least for keyboard users, but perhaps it's not hard adding keyboard navigation alongside your mouseenter/mouseleave handlers so that visually impaired users could hear the level of support + notes for engine x feature too?
2. While it's possible to figure out what JS does by interacting with page and setting breakpoints, JS code alone is becoming rather hard to follow and guess what parts of it are responsible for. Could you add some comments to toAlphabet function and the various parts of footnote logic to make future maintenance easier?

**Comment by andylizi** [comment]
Comments should be a lot clearer now. Accessibility is definitely an issue, but I'm somewhat unsure of how to go about it, mostly because of my lack of knowledge in this area. It's easy to make the cells focusable (`tabindex=0`), but it is advised that screen readers won't read `title` anyway and they shouldn't be used for accessibility……? Would something like `aria-describedby` work in this case?

Initially I've considered showing a custom tooltip on hover/focus, like the references in Wikipedia and caniuse.com. But that would involve third-party libraries and generally be a great pain to implement.

I tried highlighting the footnotes on focus, but the `blur/focusout` events doesn't seems to trigger properly for some reason……

Anyway, it's midnight for me now. I will do more research on this tomorrow.

**Comment by andylizi** [comment]
Done! I've implemented the custom tooltips and tested the accessibility with a screen reader. There should be no problem regarding that now.

By the way, I used a JS feature from ES2020, which is supported by all browsers released after 2018 or 2019. Is that acceptable for this website or should I add a fallback for it?

**Comment by RReverser** [comment]
Awesome, thanks! 

> By the way, I used a JS feature from ES2020, which is supported by all browsers released after 2018 or 2019. Is that acceptable for this website or should I add a fallback for it?

Perfectly acceptable, I don't think we should worry about older browsers on a developer-centric website.

File: roadmap.js [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+  // Also returns the HTML id of the footnote it refers to.
+  function createNoteRef(index) {
+    const id = `feature-note-${index}`;
+    return [id, h('a', { href: `#${id}` }, [`[${toAlphabet(index)}]`])];
+  }
+
+  const columnCount = 2 + Object.keys(browsers).length;
+
+  // Correspond to CSS class names.
+  const STYLE_MAP = {
```
Comment (RReverser) [inline roadmap.js:143]: As far as I can tell, this is always used together with `icon(...)` call and always to create a `<div class="feature-cell (icon class)">(text)</div>`. Let's merge them into one reusable function. Also, instead of an object, you can use `icon-${name}` template.

File: roadmap.js [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        },
+        {
+          name: 'eventListeners',
+          enabled: false
+        },
+      ]
+    });
+
+    function show() {
+      tooltip.setAttribute('data-show', '');
```
Comment (RReverser) [inline roadmap.js:320]: Is data attribute used because Popper requires it? Or can we just use a class here?

File: roadmap.js (hunk 1 above)
Comment (andylizi) [reply roadmap.js]: Good idea! It's a lot cleaner now.

**Comment by EwoutH** [comment]
This looks amazing! What's needed to merge this PR?

**Comment by RReverser** [comment]
> This looks amazing! What's needed to merge this PR?

Me having a better memory 😅

I'll try it one more time locally now and hopefully merge after that.

File: roadmap.js [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  function buildCellInner(type, text) {
+    const content = text || icon(type);
+    return h('div', { className: `feature-cell icon-${type}`}, [content]);
+  }
+
+  function renderNote(note) {
+    // No point for screen readers to pronounce those symbols out loud.
+    for (const symbol of ['✓', '✗']) {
+      if (note.startsWith(symbol)) {
+        note = `<span aria-hidden='true'>${symbol}</span>${note.substring(1)}`;
```
Comment (RReverser) [inline roadmap.js:244]: Nit: why does this one use raw HTML via `.innerHTML`, unlike the rest that use `h` API?

Comment (RReverser) [reply roadmap.js:244]: Ah, if it's because of text transform below, perhaps it can be moved around? I'd really rather avoid worrying about escaping HTML-unsafe chars :)

File: roadmap.md [hunk 4]
```diff
@@ -7,106 +7,32 @@ In November 2017, WebAssembly CG members representing four browsers, Chrome, Edg
...
-            })
-          ])
-        );
-      }
-    }
-  })();
-</script>
+<!-- Apache License 2.0, https://github.com/Remix-Design/remixicon -->
+<template id="support-symbol-yes">
+  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-label="Supported"><path class="svg-stroke" d="M10 15.2 19.2 6l1.4 1.4L10 18l-6.4-6.4L5 10.2Z"/></svg>
```
Comment (RReverser) [inline roadmap.md:12]: When I navigate to cells that just contain `✓` or `✗`, I hear e.g. "supported, blank cell" in Narrator. I think it's because those `aria-label`s live on those SVG icons, but not on the cells themselves. Worth trying to move them over?

**Comment by RReverser** [comment]
Left request for two small changes, the rest looks great!

File: roadmap.md (hunk 4 above)
Comment (andylizi) [reply roadmap.md:12]: Ah, I tested it with NVDA instead of Narrator, and it didn't have that issue. I found the exact wording can differ a lot depending on the screen reader/browser/operating system, so it's really hard to find a combination that works perfectly in all situations.

**Comment by RReverser** [comment]
Ok I think we can always iterate on further a11y improvements in separate PRs, this already looks much better than what we had. Thanks again!
